src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -241,92 +241,98 @@ def _split_heads(self, fused_qkv):
         """
         Split the last dimension into (num_heads, head_dim)
         """
-        new_tensor_shape = fused_qkv.size()[:-1] + (self.num_heads, 3 * self.head_dim)
-        # new_tensor_shape = (fused_qkv.size(1), fused_qkv.size(0)*fused_qkv.size(2), fused_qkv.size(-1))
-        # fused_qkv = fused_qkv.transpose(1, 0)
-        fused_qkv = fused_qkv.reshape(*new_tensor_shape)
-        # fused_qkv = fused_qkv.permute(0, 2, 1, 3)
-        return torch.split(fused_qkv, self.head_dim, -1)
+        batch_size, seq_length, three_times_hidden_size = fused_qkv.shape
+        fused_qkv = fused_qkv.view(batch_size, seq_length, self.num_heads, 3, self.head_dim)
+        return (fused_qkv[..., 0, :], fused_qkv[..., 1, :], fused_qkv[..., 2, :])
 
     def _merge_heads(self, x):
         # What we want to achieve is:
         # batch_size * num_heads, seq_len, head_dim -> batch_size, seq_len, num_heads * head_dim
+        batch_size_and_num_heads, seq_len, _ = x.shape
+        batch_size = batch_size_and_num_heads // self.num_heads
 
         # First view to decompose the batch size
         # batch_size*num_heads, seq_len, head_dim -> batch_size, num_heads, seq_len, head_dim
-        x = x.view(x.size(0) // self.num_heads, self.num_heads, x.size(1), self.head_dim)
+        x = x.view(batch_size, self.num_heads, seq_len, self.head_dim)
 
         # batch_size, num_heads, seq_len, head_dim -> batch_size, seq_len, num_heads, head_dim
         x = x.permute(0, 2, 1, 3)
 
         # batch_size, seq_len, num_heads, head_dim -> batch_size, seq_len, num_heads * head_dim
-        return x.reshape(x.size(0), x.size(1), self.num_heads * self.head_dim)
+        return x.reshape(batch_size, seq_len, self.num_heads * self.head_dim)
 
     def forward(
         self,
         hidden_states,
         residual,
+        alibi,
+        attention_mask,
         layer_past=None,
-        attention_mask=None,
-        alibi=None,
         head_mask=None,
         use_cache=False,
         output_attentions=False,
     ):
-        alibi = alibi.to(hidden_states.device)  # to make the model possible to run under accelerate
```

Comment:
Yes, as long as it's an input the layer, it'll be on the right device.

Reply:
@sgugger Could you elaborate this a bit for me 😄 🙏 if not too busy.

Reply:
The hooks that Accelerate adds will place each input received by each submodule on the proper device just before the forward. Since `alibi` is an input of this module, it will be placed on the device alongside `hidden_states`, so this line is un-necessary for `Accelerate`.
(Note that I'm trusting the comment that this was only necessary for Accelerate model parallelism purposes.)

Reply:
(I'm also trusting the comment)

Reply:
Great thanks for the detailed explanation @sgugger !